### PR TITLE
Module quotients

### PR DESCRIPTION
**Pull request by abooij**
Given a submodule of a module over a ring, construct its quotient module.

This is work in progress.

### CODE REVIEW

**Comment by siddharthist** [comment]
It looks like this should be merged after #831, is that right?

**Comment by abooij** [comment]
Yes, this is based on #831.

Currently this PR adds the construction of a quotient module. I will also add the universal property.

The code is somewhat repetitive right now.

**Comment by abooij** [comment]
@fpvandoorn I added the quotient map and the universal property.

File: UniMath/.dir-locals.el [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
+  . ((eval .
 	   (let ((unimath-topdir (expand-file-name (locate-dominating-file buffer-file-name "UniMath"))))
 	     (setq fill-column 100)
 	     (make-local-variable 'coq-use-project-file)
 	     (setq coq-use-project-file nil)
 	     (make-local-variable 'coq-prog-args)
 	     (setq coq-prog-args `("-emacs" "-indices-matter" "-type-in-type" "-w" "-notation-overridden,-local-declaration,+uniform-inheritance,-deprecated-option" "-Q" ,(concat unimath-topdir "UniMath") "UniMath" ))
 	     (make-local-variable 'coq-prog-name)
-	     (setq coq-prog-name (concat unimath-topdir "sub/coq/bin/coqtop"))
+;	     (setq coq-prog-name (concat unimath-topdir "sub/coq/bin/coqtop"))
```
Comment (siddharthist) [inline UniMath/.dir-locals.el]: These changes should not be checked in.

Comment (abooij) [reply UniMath/.dir-locals.el]: Oops, thanks.

**Comment by DanGrayson** [comment]
Ideally, a PR that depends on another PR would be submitted only after the first one has been accepted, for otherwise we end up reviewing the changes of the first PR twice.

**Comment by abooij** [comment]
@DanGrayson, I submitted this PR so that people know where to find this work. Once it is ready for review (indeed, ideally after merging of the PR that this depends on) I will update this PR.

**Comment by DanGrayson** [comment]
It's easy to find closed unmerged pull requests by a given author, e.g.: https://github.com/UniMath/UniMath/pulls?q=is%3Apr+is%3Aclosed+is%3Aunmerged+author%3ADanGrayson

So shall we close this until it is ready for review?

**Comment by abooij** [comment]
The only work remaining for this is rebasing on top of #831.

**Comment by abooij** [comment]
Rebased, this is now ready for review.

File: UniMath/Algebra/Modules/Quotient.v [hunk 2]
```diff
@@ -0,0 +1,205 @@
+Require Import UniMath.Foundations.Sets.
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: Could you please add a header to the file saying what it contains, who wrote it and when. You can take a look at some of the other files for inspiration.

File: UniMath/Algebra/Modules/Quotient.v [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+Require Import UniMath.Algebra.Modules.Submodule.
+
+Section quotmod_rel.
+
+  Context {R : rng}
+          (M : module R).
+
+
+  Local Open Scope module_scope.
+  Definition isacthrel (E : hrel M) : UU :=
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: I guess this is short for `isactionhrel`? I don't think it would hurt to spell it out or even call it `islactionhrel` to clarify that the action is on the left. I know that many of the identifiers in Algebra are extremely short and that there is no global naming convention yet, but I would personally prefer a slightly longer name here so that I don't have to guess what it means.

File: UniMath/Algebra/Modules/Quotient.v [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        simpl in H; rewrite H.
+        use eqrelrefl.
+      * unfold unel, quotmod_rngact; simpl; rewrite (setquotunivcomm E).
+        apply maponpaths.
+        apply module_mult_1.
+  Defined.
+
+  Definition quotmod_rngfun : rngfun R (rngofendabgr quotmod_abgr).
+  Proof.
+    unfold rngfun. unfold rigfun.
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: You can merge this into one unfold by `unfold rngfin, rigfun.`

File: UniMath/Algebra/Modules/Quotient.v [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    use rigfunconstr.
+    - use quotmod_rngmap.
+    - use mk_isrigfun;
+        [ use mk_ismonoidfun | use mk_ismonoidfun ];
+        [ use mk_isbinopfun; intros r r' | | use mk_isbinopfun; intros r r' | ].
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: I'm not sure exactly what you're constructing here, but it seems like you have to have to dig very deep to get the goals that you need. I have a feeling that there is a constructor lemma missing?

File: UniMath/Algebra/Modules/Quotient.v [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+      + use module_mult_0_to_0.
+      + use module_mult_assoc.
+      + use module_mult_unel2.
+  Defined.
+
+  Definition quotmod_mod_struct : module_struct R quotmod_abgr.
+  Proof.
+    unfold module_struct.
+    exact quotmod_rngfun.
+  Defined.
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: Wouldn't it be clearer to write:
```
Definition quotmod_mod_struct : module_struct R quotmod_abgr := quotmod_rngfun.
```
?

File: UniMath/Algebra/Modules/Quotient.v [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+  Definition quotmod : module R.
+  Proof.
+    use modulepair.
+    - exact quotmod_abgr.
+    - exact quotmod_mod_struct.
+  Defined.
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: How about having a notation for this like `M / A`? (or if that doesn't work `M %/ A`)

File: UniMath/Algebra/Modules/Quotient.v [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+  Definition quotmod_quotmap : modulefun M quotmod.
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: It would be nice to have a notation for `modulefun` as well. Maybe something like `Mod(M,N)` could work? One could alternatively use some fancy arrow, but I think I prefer to see which category we are in just by seeing the notation.

File: UniMath/Algebra/Modules/Quotient.v [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+      + use module_mult_unel2.
+  Defined.
+
+  Definition quotmod_mod_struct : module_struct R quotmod_abgr.
+  Proof.
+    unfold module_struct.
+    exact quotmod_rngfun.
+  Defined.
+
+  Definition quotmod : module R.
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: Having the notation `R-mod` for `module R` could look nice

File: UniMath/Algebra/Modules/Quotient.v [hunk 10]
```diff
@@ -0,0 +1,205 @@
+Require Import UniMath.Foundations.Sets.
+Require Import UniMath.MoreFoundations.All.
+Require Import UniMath.Algebra.Rigs_and_Rings.
+Require Import UniMath.Algebra.Monoids_and_Groups.
+Require Import UniMath.Algebra.Modules.Core.
+Require Import UniMath.Algebra.Modules.Submodule.
+
+Section quotmod_rel.
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: Could you add a comment starting with `(** * ` before the `Section` saying what it contains? By formatting the comment like this it will display as a section header in the HTML documentation

File: UniMath/Algebra/Modules/Quotient.v [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+
+  Definition isacthrelmonoideqrel (E : monoideqrel) : isacthrel E := pr2 (pr2 E).
+  Coercion isacthrelmonoideqrel : monoideqrel >-> isacthrel.
+
+  Definition mk_monoideqrel (E : eqrel M) : isbinophrel E -> isacthrel E -> monoideqrel :=
+    λ H0 H1, (E,,(H0,,H1)).
+
+End quotmod_rel.
+
+Section quotmod_submodule.
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: Please add a comment here as well saying what the section contains

File: UniMath/Algebra/Modules/Quotient.v [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+      assert (H : (r * m) - (r * m') = r * (m - m')) by
+          now rewrite module_inv_mult, module_mult_is_ldistr.
+      generalize (submodulemult A r (m - m')).
+      simpl in *.
+      now rewrite H.
+  Defined.
+
+End quotmod_submodule.
+
+Section quotmod_def.
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: And a comment here as well please

File: UniMath/Algebra/Modules/Quotient.v [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+  Context {R : rng}
+          (M : module R).
+
+
+  Local Open Scope module_scope.
+  Definition isacthrel (E : hrel M) : UU :=
+    ∏ r a b, E a b -> E (r * a) (r * b).
+
+  Definition monoideqrel : UU :=
+    total2 (λ E : eqrel M, isbinophrel E × isacthrel E).
```
Comment (mortberg) [inline UniMath/Algebra/Modules/Quotient.v]: You can use `\sum` instead of `total2` here in order for it to display nicely as a Sigma

File: UniMath/Algebra/Modules/Quotient.v (hunk 8 above)
Comment (mortberg) [reply UniMath/Algebra/Modules/Quotient.v]: Thinking a bit more about it I realized that one might want to see the ring R in the notation as well. So maybe `R-mod(M,N)` or `Mod_R(M,N)` could work better

File: UniMath/Algebra/Modules/Quotient.v (hunk 5 above)
Comment (abooij) [reply UniMath/Algebra/Modules/Quotient.v]: Are you talking specifcally about this constructor unfolding here, or also the lines starting with `all:` following it?

File: UniMath/Algebra/Modules/Quotient.v (hunk 7 above)
Comment (abooij) [reply UniMath/Algebra/Modules/Quotient.v]: I can add that. However, could you point me to a good example of how and where notation should be specified? I'm a little bit worried I will cause trouble for other people down the road. (I am also thinking about the discussion we had in Birmingham in December.)

File: UniMath/Algebra/Modules/Quotient.v (hunk 5 above)
Comment (mortberg) [reply UniMath/Algebra/Modules/Quotient.v]: I'm talking about all of the uses of `use`

File: UniMath/Algebra/Modules/Quotient.v (hunk 7 above)
Comment (mortberg) [reply UniMath/Algebra/Modules/Quotient.v]: Something along the lines of:

```
Local Notation "M / A" := (quotmod M A) : module_scope.
```

after closing the section should do the trick.  You might have to specify level and associativity as well (they are Reserved at level 40 with left associativity in the Coq standard library).  Every file that depends on this file in which you want to use this notation will have to redefine this notation if you do it this way, this is a bit verbose but it makes it possible for us to avoid the kind of issues that we had in Birmingham.

File: UniMath/Algebra/Modules/Quotient.v (hunk 5 above)
Comment (abooij) [reply UniMath/Algebra/Modules/Quotient.v]: I have written this ever so slightly differently, and added explanations in English. Does the new version still need a new constructor lemma?

File: UniMath/Algebra/Modules/Quotient.v (hunk 7 above)
Comment (abooij) [reply UniMath/Algebra/Modules/Quotient.v]: But if every file that depends on this file needs to re-introduce the notation, then adding it after the end of this section won't do much, as that's (currently) the end of the file. Or am I missing something?

Comment (mortberg) [reply UniMath/Algebra/Modules/Quotient.v]: Well, now that you have the definition there should ideally be some theory about the construction in a section at the end of the file. This theory could use this notation. Furthermore, if you have the notation declared it will be easier for someone else to copy-paste it. :)

File: UniMath/Algebra/Modules/Quotient.v (hunk 5 above)
Comment (mortberg) [reply UniMath/Algebra/Modules/Quotient.v]: No, I like the new version!

File: UniMath/Algebra/Modules/Quotient.v (hunk 8 above)
Comment (abooij) [reply UniMath/Algebra/Modules/Quotient.v]: @mortberg, I have added some notation commits. Is this what you had in mind?

Comment (mortberg) [reply UniMath/Algebra/Modules/Quotient.v]: Yes, excellent!

**Comment by DanGrayson** [comment]
@abooij : re: " I submitted this PR so that people know where to find this work. "

The ideal way to advertise work is as a link in a comment on the issue that solicits the work.  Also, being assigned to an issue is a good way to advertise who is working on the issue.

**Comment by abooij** [comment]
@DanGrayson This work was mergable, compiled, and was beneficial to the project, at PR time. So, following the style of many other open source projects, I submitted this work as a PR with a WIP tag, to both indicate that the work was at an advanced state, and to allow others to voice their opinion. I am sorry if I broke any of the contribution guidelines of this project.

**Comment by benediktahrens** [comment]
On 01/05/2018 11:51 AM, Daniel R. Grayson wrote:
> @abooij : re: " I submitted this PR so that people know where to find
> this work. "
> 
> The ideal way to advertise work is as a link in a comment on the
> issue that solicits the work.  Also, being assigned to an issue is a
> good way to advertise who is working on the issue.
> 

Not every work is solicited, or announced as an issue, so what you 
consider ideal is not always possible.

I would say that issues announcing future work are much more of a 
nuisance than a PR that is opened not to be merged, but to be discussed.
The reason is that announcing or soliciting work in an issue might 
happen without any work ever being done, that is, it is noise without 
benefit.
In a PR, as @abooij did, some work has been done already, so even if the 
noise is a nuisance, at least it comes with evidence of some achieved work.
In summary, I would like to discourage people from announcing their 
plans in issues.


One facility that would help is allowing PR submitters to add tags to 
their PRs. Unfortunately, only contributors with write access can modify 
tags.
Until that changes, I think that @abooij's way of announcing their work 
is the best one available.

**Comment by DanGrayson** [comment]
The file `Algebra/.package/files` should list Modules/Submodule.v only once.

**Comment by DanGrayson** [comment]
Why not make the notations global?  That was others can use them without copying the code -- all they have to do is to open the appropriate scope.  It won't be annoying to those who don't want to have the notations, as the scope will not be open.

**Comment by DanGrayson** [comment]
A minor comment: It seems a bit unusual to adopt category theory notation such as 
```
Local Notation "R-mod( M , N )" := (modulefun M N) : module_scope.
```
without actually having any categories around, and it doesn't save much clutter, either.

**Comment by DanGrayson** [comment]
Introducing that notation also will tend to make the origin of lemma names more opaque to the reader, such as this one:
``` 
-Lemma isasetmodulefun {R : rng} (M N : module R) : isaset (modulefun M N).
+Lemma isasetmodulefun {R : rng} (M N : module R) : isaset (R-mod(M, N)).
```

**Comment by DanGrayson** [comment]
"Quotient" is a noun, not a verb, so this comment has a grammatical error:
```
Quotienting a module over a ring by a submodule
```

**Comment by DanGrayson** [comment]
Some of the arguments seem a bit convoluted, because of the frequent use of `generalize`.  Compare the way the proof steps go, for example, before and after this change:
```
diff --git a/UniMath/Algebra/Modules/Quotient.v b/UniMath/Algebra/Modules/Quotient.v
index 74714bb..683a32d 100644
--- a/UniMath/Algebra/Modules/Quotient.v
+++ b/UniMath/Algebra/Modules/Quotient.v
@@ -59,8 +59,8 @@ Section quotmod_submodule.
       generalize (submoduleadd A (x - y) (y - z) xy yz).
       now rewrite (assocax M), <- (assocax M (grinv _ y) y), grlinvax, lunax.
     - intros x.
-      generalize (submodule0 A).
-      now rewrite <- (grrinvax M x).
+      rewrite grrinvax.
+      apply submodule0.
     - intros x y axy.
       generalize (submoduleinv A (x - y) axy).
       now rewrite grinvop, grinvinv.
```
Doesn't that make it clearer?

**Comment by DanGrayson** [comment]
It might be a bit clearer to use `exact` instead of `use` when there are no arguments to be filled in, as here:
```
diff --git a/UniMath/Algebra/Modules/Quotient.v b/UniMath/Algebra/Modules/Quotient.v
index 74714bb..5a6f5b1 100644
--- a/UniMath/Algebra/Modules/Quotient.v
+++ b/UniMath/Algebra/Modules/Quotient.v
@@ -149,7 +149,7 @@ Section quotmod_def.
   Proof.
     unfold rngfun, rigfun.
     use rigfunconstr.
-    - use quotmod_rngmap.
+    - exact quotmod_rngmap.
     - use mk_isrigfun.
       (* To show that quotmod_rngmap is a ring action, we show it is a monoid homomorphism with
       respect to both monoids on R. *)
```

**Comment by DanGrayson** [comment]
I have no further comments.

**Comment by abooij** [comment]
Re: applications of `generalize`

In fact, it may not be trivial to get rid of some of these, because Coq won't be able to do the relevant `rewrite`s otherwise. I've gotten rid of the one you mentioned, but am having more trouble with the other ones. This is the technique that we worked out in Birmingham, and that seemd to be most reliable for doing `rewrite`-based proofs.

**Comment by abooij** [comment]
I believe I have processed your comments.

File: UniMath/Algebra/Modules/Quotient.v [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+
+  Context {R : rng}
+          (M : module R).
+
+
+  Local Open Scope module_scope.
+  Definition isactionhrel (E : hrel M) : UU :=
+    ∏ r a b, E a b -> E (r * a) (r * b).
+
+  Definition monoideqrel : UU :=
```
Comment (DanGrayson) [inline UniMath/Algebra/Modules/Quotient.v]: The name `monoideqrel` suggests that you are defining an equivalence relation on a monoid, but
actually, you have a relation on a module that is compatible with addition and with scalar multiplication.
Could you choose a better name?  What about `module_eqrel`?

File: UniMath/Algebra/Modules/Quotient.v [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+  Local Notation "x + y" := (@op _ x y).
+  Local Notation "x - y" := (@op _ x (grinv _ y)).
+  Local Open Scope module_scope.
+
+  Definition eqrelsubmodule : eqrel M.
+  Proof.
+    use eqrelconstr.
+    - exact (λ m m', A (m - m')).
+    - intros x y z xy yz.
+      generalize (submoduleadd A (x - y) (y - z) xy yz).
```
Comment (DanGrayson) [inline UniMath/Algebra/Modules/Quotient.v]: To remove this use of "generalize", resulting in something more understandable, one may do this:
```
     use eqrelconstr.
     - exact (λ m m', A (m - m')).
     - intros x y z xy yz.
-      generalize (submoduleadd A (x - y) (y - z) xy yz).
-      now rewrite (assocax M), <- (assocax M (grinv _ y) y), grlinvax, lunax.
+      assert (K := submoduleadd A (x - y) (y - z) xy yz); clear xy yz.
+      rewrite (assocax M) in K.
+      rewrite <- (assocax M _ y) in K.
+      rewrite grlinvax in K.
+      rewrite lunax in K.
+      exact K.
     - intros x.
       rewrite (grrinvax M x).
       exact (submodule0 A).
```
It is also a bit more instructive to do one rewrite at a time, so the student can see what's happening.

Other uses of "generalize" can be handled the same way, as well as the places where you introduce all but one variable.

Comment (DanGrayson) [reply UniMath/Algebra/Modules/Quotient.v]: And the proof above looks better when stepping through it if some more notations are added:
```
 
   Local Notation "x + y" := (@op _ x y).
   Local Notation "x - y" := (@op _ x (grinv _ y)).
+  Local Notation "  - y" := (grinv _ y).
+  Local Notation "0" := (unel _).
   Local Open Scope module_scope.
 
   Definition eqrelsubmodule : eqrel M.
```

File: UniMath/Algebra/Modules/Quotient.v [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+    use modulefunpair.
+    - exact (setquotpr E).
+    - now use (ismodulefunpair (mk_isbinopfun _)).
+  Defined.
+
+  Definition quotmoduniv
+             (N : module R)
+             (f : R-mod(M, N))
+             (is : iscomprelfun E f) :
+    R-mod(quotmod, N).
```
Comment (DanGrayson) [inline UniMath/Algebra/Modules/Quotient.v]: This is a fine way to phrase the universal property of the quotient module, but what about including also the standard version, where the quotient module is formed from a module and a submodule?

File: UniMath/Algebra/Modules/Submodule.v [hunk 17]
```diff
@@ -124,3 +124,28 @@ Defined.
...
+          {M : module R}
+          (A : submodule M).
+
+  Local Notation "x + y" := (@op _ x y).
+  Local Notation "x - y" := (@op _ x (grinv _ y)).
+
+  Definition submoduleadd (x y : M) : A x -> A y -> A (x + y).
+  Proof.
+    intros ax ay.
+    exact (pr1 (pr1 (pr1 (pr2 A))) (carrierpair A x ax) (carrierpair A y ay)).
```
[the remark this answers is not in the document]
Comment (abooij) [reply UniMath/Algebra/Modules/Submodule.v]: These are the accessor functions.

File: UniMath/Algebra/Modules/Quotient.v (hunk 14 above)
Comment (abooij) [reply UniMath/Algebra/Modules/Quotient.v]: Ah yeah, that's a typo, I'll fix it.

File: UniMath/Algebra/Modules/Quotient.v (hunk 16 above)
Comment (abooij) [reply UniMath/Algebra/Modules/Quotient.v]: Done.

File: UniMath/Algebra/Modules/Submodule.v (hunk 17 above)
[the remark this answers is not in the document]
Comment (DanGrayson) [reply UniMath/Algebra/Modules/Submodule.v]: Not exactly -- an accessor function would consist of just a string of projections.

[the remark this answers is not in the document]
Comment (abooij) [reply UniMath/Algebra/Modules/Submodule.v]: Wait, what is the difference between a "chain" of projections and a "string" of projections?

[the remark this answers is not in the document]
Comment (DanGrayson) [reply UniMath/Algebra/Modules/Submodule.v]: No difference.  But you also have "carrierpair" getting used in this function, which is not a projection.

[the remark this answers is not in the document]
Comment (abooij) [reply UniMath/Algebra/Modules/Submodule.v]: Okay. Feel free to take this PR and fix it. I am tired of guessing what the exact code style guide of this project is.

[the remark this answers is not in the document]
Comment (DanGrayson) [reply UniMath/Algebra/Modules/Submodule.v]: I'm sorry -- perhaps we should loosen the style guide at that point.  The current statement is
```
* Define and use accessor functions for structures instead of chains
  of `pr1` and `pr2`. This makes the code easier to maintain in the
  long run (if the structure is rearranged the proofs will still work
  if the accessor functions are changed accordingly).
```
Your function doesn't deviate much from that...

**Comment by DanGrayson** [comment]
It looks good to me.  The only conflicts are of the following form:
```
<<<<<<< HEAD
  Context {R : ring} {M N : module R} (f : modulefun M N).
||||||| merged common ancestors
  Context {R : rng} {M N : module R} (f : modulefun M N).
=======
  Context {R : rng} {M N : module R} (f : R-mod(M, N)).
>>>>>>> 8e15408f
```
so they should be easy to fix.  Can you do that, Auke?

**Comment by abooij** [comment]
The conflict is a result of the request to change notation in a [previous comment](https://github.com/UniMath/UniMath/pull/833#issuecomment-356444959).

I attempted to rebase, but likely there have been other changes since that I've missed. Let's see what CI says.

**Comment by abooij** [comment]
The CI now passes.
